yt/frontends/gizmo/data_structures.py

```diff
@@ -49,3 +54,113 @@ def _is_valid(cls, filename, *args, **kwargs):
+                self.omega_radiation = hvals["Omega_Radiation"]
+                self.hubble_constant = hvals["HubbleParam"]
+            else:
+                # Should still support old GIZMO versions too
+                only_on_root(
+                    mylog.info,
+                    "ComovingIntegrationOn does not exist, falling back to OmegaLambda",
+                )
+                self.omega_lambda = hvals["OmegaLambda"]
+                self.omega_matter = hvals["Omega0"]
+                self.hubble_constant = hvals["HubbleParam"]
+                self.cosmological_simulation = self.omega_lambda == 0.0
+        except KeyError:
+            # If these are not set it is definitely not a cosmological dataset.
+            self.omega_lambda = 0.0
+            self.omega_matter = 1.0  # Just in case somebody asks for it.
+            self.cosmological_simulation = 0
+            # Hubble is set below for Omega Lambda = 0.
+
+        # Since new versions of GIZMO output the ComovingIntegrationOn flag
+        # we will rely on its presence to determine if this is a cosmological
+        # dataset.
+        if not self.cosmological_simulation:
+            only_on_root(
+                mylog.info,
+                "ComovingIntegrationOn != 1 or (not found "
+                "and OmegaLambda is 0.0), so we are turning off Cosmology.",
+            )
+            self.hubble_constant = 1.0  # So that scaling comes out correct
+            self.cosmological_simulation = 0
+            self.current_redshift = 0.0
+            # This may not be correct.
+            self.current_time = hvals["Time"]
+        else:
+            # Now we calculate our time based on the cosmology, because in
+            # ComovingIntegration hvals["Time"] will in fact be the expansion
+            # factor, not the actual integration time, so we re-calculate
+            # global time from our Cosmology.
+            cosmo = Cosmology(
+                hubble_constant=self.hubble_constant,
+                omega_matter=self.omega_matter,
+                omega_lambda=self.omega_lambda,
+            )
+            self.current_time = cosmo.lookback_time(self.current_redshift, 1e6)
+            only_on_root(
+                mylog.info,
+                "Calculating time from %0.3e to be %0.3e seconds",
+                hvals["Time"],
+                self.current_time,
+            )
+        self.parameters = hvals
+
+        prefix = os.path.abspath(
+            os.path.join(
+                os.path.dirname(self.parameter_filename),
+                os.path.basename(self.parameter_filename).split(".", 1)[0],
+            )
+        )
+
+        if hvals["NumFiles"] > 1:
+            for t in (
+                f"{prefix}.%(num)s{self._suffix}",
+                f"{prefix}.gad.%(num)s{self._suffix}",
+            ):
+                if os.path.isfile(t % {"num": 0}):
+                    self.filename_template = t
+                    break
```

Comment:
the `for/else` construct might be obscure to some readers. I'm not sure it is justified here, given that this is equivalent to a simple `if/elif/else`

Reply:
Just fyi, you're the one who set up that structure :laughing: (#3899). Since the point of this was to add Gadget 2 file template support, I'm inclined to just revert the original change.